### PR TITLE
feat: implement CrimeKNN brute-force classifier with combined distance

### LINKED ISSUE #66: US-2.4: KNN Prediction Algorithm Story Points: 8
Acceptance Criteria:
• CrimeKNN class with predict(query) method in src/algorithms/knn/knn.py
• Query accepts: {latitude, longitude, hour, day_of_week, day_of_year, primary_type}
• Training data pre-filtered by primary_type before KD-tree construction
• Prediction = inverse-distance weighted average of arrest outcome (0 or 1) among K neighbours
• Formula: sum(arrest_i / dist_i) / sum(1 / dist_i) — returns float in [0, 1]
• Train/test split: 80% train (2001–2022), 20% test (2023–2025)
• Accuracy metric: MAE reported on held-out test set
• K parameter tunable (test K = 5, 10, 20); optimal K documented
• Edge case: crime type with fewer records than K returns available neighbours only
• Edge case: distance of 0 handled (use small epsilon to avoid division by zero)

Tasks:

- [ ] 1. Create CrimeKNN class in src/algorithms/knn/knn.py
- [ ] 2. Implement __init__: accepts records + k + max_spatial_km, builds KDTree
- [ ] 3. Implement predict: filter by primary_type, query KD-tree, compute weighted arrest rate
- [ ] 4. Add epsilon guard for zero-distance neighbours (distance = max(dist, 1e-6))
- [ ] 5. Create train/test split by date (cutoff: January 1 2023)
- [ ] 6. Generate predictions for test set
- [ ] 7. Calculate MAE: mean(|predicted_rate - actual_arrest|) across test records
- [ ] 8. Tune K: test K = 5, 10, 20; report MAE for each
- [ ] 9. Document optimal K choice in code comments
- [ ] 10. Save test predictions to CSV for comparison dashboard

### PR DESCRIPTION
**Pull request by carl-harper**
## What does this PR do?
It builds a knn from scratch that integrates the combined distance formula that we built before. It is not really useable for KD because haversine distance isn't, and haversine (non-euclidian) is part of the cobined distance formula (other half is cyclical temporal for month, time etc). 
## How can a reviewer test this?
needs to be tested for functionalty, do the elbow diagram test. I did not do some of the things in the checklist. Check
(if out data is not hte same)
Option 1 — Unit test with synthetic data (easiest)
Option 2 — Notebook demo
A small Jupyter notebook that walks through:
pythonfrom src.algorithms.knn.knn import CrimeKNN, split_by_date, tune_k
# loads sample data, fits, predicts, saves CSV

otherwise: 
python run_knn.py

## Any edge cases to be aware of?
end of the year, end of the day, etc

## Checklist
- [ ] Tests written
- [ ] No direct push to main
- [ ] Reviewed by at least one teammate

### CODE REVIEW

**Comment by ght-1** [comment]
File seems empty :(

**Comment by ght-1** [comment]
We need to adapt this to make it work with the dataframe as per the last within-estimation cleaning decisions.
